Support Option<Vec> and Option<String>

kuecks opened the pull request:
The main issue with supporting these structs is that they are both 3 pointers wide, while the existing code relied on Option<T> being precisely 1 pointer wide (and thus only worked for pointer types). To resolve this, this PR has 2 commits:

1) Instead of using a single RustOption<T> type to represent the C++ binding, using traits I defined different structs for the FFI representation of Option<T> depending on T. For the old pointer types like &T and Box<T> this is the old RustOption<T>, but this now allows me to define new types (with a different size) to use for Option<Vec> and Option<String>

2) Actually implement Option<Vec> and Option<String> using the indirection above

kuecks commented on src/lib.rs:
```diff
@@ -513,8 +512,8 @@ pub mod private {
-    pub use crate::rust_option::assert_option_safe;
-    pub use crate::rust_option::RustOption;
+    pub use crate::rust_option::OptionFfi;
+    pub use crate::rust_option::OptionFfiInverse;
```
so the struct RustOption no longer exists. Instead we now use the trait `OptionFfi` (implemented on `std::option::Option`) to link an FFI partner struct for Option<T> (for some T). This way the partner struct can be completely different (notably a different size) for different T. `OptionFfiInverse` is implemented on the partner struct as a helper to convert back to `std::option::Option`

kuecks commented on src/rust_option.rs:
```diff
@@ -19,390 +19,552 @@ use core::pin::Pin;
-    }
+// Pointer-sized pointer types with niche optimization
+const _: () = {
+    impl_option_ffi! { <T: OptionPtrTarget>, usize, &'static () }
```
Here I define the partner struct (what was RustOption). So that I don't need a unique name for each different type, the definition is tucked away in a const. This means that we can't name this type outside of this block. The rest of the commit is removing all references to `RustOption` outside and replacing it with `<Option<T> as OptionFfi>::Target` to deal with this

kuecks commented on include/cxx.h:
```diff
@@ -383,91 +476,68 @@ class Option<Box<T>> {
 };
-#endif // CXXBRIDGE1_RUST_FN
+
+OPTION_DIRECT_DECL(String, std::uintptr_t, 3)
```
Looks a little silly to use a macro for just 1 impl, but this will be reused in a follow up when I support Option<primitive>

capickett commented on gen/src/write.rs:
```diff
@@ -1114,15 +1148,21 @@ fn write_rust_function_shim_impl(
-        ) = &sig.ret
-        {
-            write!(out, ")");
+        if let Some(ret) = &sig.ret {
```
This is starting to look like it wants to be a match instead of if-let chain

capickett commented on src/rust_option.rs:
```diff
@@ -19,390 +19,594 @@ use core::pin::Pin;
+        }
+
+        /// SAFETY: ptr must be valid for 'a
+        pub unsafe fn from_raw_double_repr(
```
What does "raw double repr" mean?

kuecks replied:
`repr` is for `*RustVec<T> -> Option<&Vec<T>>` and `double_repr` is specifically for `
*RustVec<RustString>` -> `Option<&Vec<String>>` (both the container and the inner type change)

However, I can also just use `as _` magic casts on the pointers to get rid of both the repr and double_repr methods (just `from_raw`) so I think I'm gonna do that instead

capickett reviewed (commented):
Does `is_direct` correspond with `needs_indirect_abi`? Or are they unrelated? I was a bit confused by the naming there.

How about `Option<&str>`. Is that supported here? I see impl for `Option<&String>` but not str.

kuecks commented:
> Does `is_direct` correspond with `needs_indirect_abi`? Or are they unrelated? I was a bit confused by the naming there.

Hrmm is_direct was referring to the fact that `set` took in `T` instead of `T*` but now I'm not sure why I did that. It seems like I don't need to make that distinction, but now I'm not entirely sure how the code was working...

I think that the code was actually magically passing T* anyway due to calling convention so my code was luckily working. I removed the `is_direct` and always pass a pointer which seems more correct now (the spooky thing was that I changed the signature on the C++ side to say I pass a pointer over FFI instead of a value, but didn't change the rust side that received it. The rust side had always expected a pointer so something was strange before to say the least)

> How about `Option<&str>`. Is that supported here? I see impl for `Option<&String>` but not str.

Option<&str> is not yet implemented because it is a fat pointer. Planning on doing it in a future diff though
